### ISSUE TITLE
PEP 3113 Remove use of tuple parameter unpacking

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8346]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)| @rodrigc reported|
|-|-|
|Trac ID|trac#8346|
|Type|enhancement|
|Created|2016-05-22 09:03:03Z|
|Branch|https://github.com/twisted/twisted/tree/8346-rodrigc-pep3113|

PEP 3113 ( ​https://www.python.org/dev/peps/pep-3113/ ) deprecates tuple parameter unpacking.

This syntax:
```
def fxn(a, (b, c), d):
    pass
```

should be changed to:

```
def fxn(a, b_c, d):
    b, c = b_c
    pass
```

Both syntaxes work in Python 2.6 and Python 2.7.  Only the new syntax works in Python 3.x

<details><summary>Searchable metadata</summary>

```
trac-id__8346 8346
type__enhancement enhancement
reporter__rodrigc rodrigc
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8346_rodrigc_pep3113 8346-rodrigc-pep3113
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1463907783589907 1463907783589907
changetime__1474198049187875 1474198049187875
version__None None
owner__Amber_Brown__HawkOwl___hawkowl_____ Amber Brown (HawkOwl) <hawkowl@...>

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_1"></a>|@rodrigc commented|
|-|-|

See:
https://github.com/twisted/twisted/pull/67

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @rodrigc|
|-|-|

Thanks for the changes.

They look good, but there is a problem with this automated conversion, as it produces argument names which don't comply with Twisted coding convention.

For example in conch/ssh/forwarding.py `packOpen_direct_tcpip(conn_host_port, orig_host_port)` this is public API and we should have a name for the arguments which comply with our coding standards.

So maybe they argument should be named packOpen_direct_tcpip(destination, source) and document what value should have the destionation and the source.

A lot of the arguments used by callback functions can just be called `result`.

I will have to get more feedback from Glyph and Amber as the wiki page dedicated to py3 porting specifies that 2to3 tools should not be used https://twistedmatrix.com/trac/wiki/Plan/Python3

now I don't know why 2to3 should not be used... and I think that is fine to use it as a helper :) 

... but yes, the output from 2to3 should be manually reviewd and updated to comply with Twisted coding standards

We can not land this branch as it is.

Will you have time to update the code to comply with the coding standard?

This branch can also be broken into smaller branches so that we have e smaller diff.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_4"></a>|@rodrigc commented|
|-|-|

Regarding 2to3, glyph wrote that comment here: https://twistedmatrix.com/trac/wiki/Plan/Python3?version=54
so getting his feedback would be useful.  In the past, 2to3 was not so robust, and had more problems than it does now.

However, I don't agree with that comment in its current form.

What I would agree with is this:

1.  Running 2to3 and just submitting one big patch without analysis or testing is **NOT OK**.
1.  Running smaller subsets of 2to3 by doing **2to3 -f [fixer]**, and performing better analysis and testing when submitting patches is **OK**.

I've been trying to break things into smaller patches with **2to3 -f** as I have been submitting patches.

Some of these Python3 tickets have been sitting for 4 years, so it would be good to pick up the pace and make more progress on them.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_5"></a>|@rodrigc commented|
|-|-|

When I ran **2to3 -f tuple_params**, the tool provides output like this:

```
--- ./docs/core/examples/cred.py        (original)
+++ ./docs/core/examples/cred.py        (refactored)
@@ -106,7 +106,8 @@
         self.sendLine("You have the following privileges: ")
         self.sendLine(" ".join(map(str, self.avatar.getPrivileges())))
 
-    def _cbLogin(self, (interface, avatar, logout)):
+    def _cbLogin(self, xxx_todo_changeme):
+        (interface, avatar, logout) = xxx_todo_changeme
         assert interface is IProtocolUser
         self.avatar = avatar
         self.logout = logout
```

In my patch submitted to https://github.com/twisted/twisted/pull/67, I did **not** take this output directly and changed **xxx_todo_changeme** to **interface_avatar_logout**.
It wasn't clear to me that this name was against the coding standard: http://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html#naming

I chose that name to keep it as close as possible to the existing code.

Same with the other names I chose.

If you have better name suggestions, I am OK with changing the patch, but I need more
detailed feedback on what names to choose.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

Many thanks for the feedback.

I think that is ok to use to use the 2to3 tools as a start.

So the Coding conversion is not explicit about the argument names... but I think that the whole Twisted code is against the underline as a separator.

I think that the names you chose are very good ... as a start.

I need to clarify if underlines are allowed in the argument names and if using composed names for tuple arguments is also allowed by the coding standard.

Somehow the composed arguments names are strange to me... but maybe only because there are to good :)

Will come back with more feedback. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_7"></a>|@rodrigc set owner to @adiroiban|
|-|-|

OK, waiting for more feedback.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

some feedback from Glyph (BST time)

```
(22:23:05) glyph: adiroiban: argument names are variables, just like any other kind of variable
(22:23:13) glyph: adiroiban: so, initial lower case, camelcase
```

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set owner to @rodrigc|
|-|-|

rodrigc thanks for the work on this.

I left a few suggestion in the PR.

My main comments are:

1. argument names should be initial lower case and then camelcase
2. For the public API we should add minimal documentation as the arguments are no longer self documented.
3. For public callback methods, just use `result`
4. For private callback methods, use `result` or any name that you like as long as it is compatible with the Twisted coding convention
5. For the public API try to give the a name based on their function .. like geometry, position, dimension, peer

--------

Does this make sense?

Can you use this feedback to update the branch?

We don't name many public places in which arguments are used in public API.
Also, in previous public API tuple arguments had no names so we can use any name.

I would like to have names in standard Twisted format so that we don't generate new warnings in twistedchecker

Many thanks!

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_10"></a>|@rodrigc set owner to @adiroiban|
|-|-|

I incorporated your feedback and uploaded a new patch:

https://github.com/twisted/twisted/pull/67

Please review

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

Changes look good.

Your branch was still having variable names which were not compliant with the Twisted coding standard.

I have updated them while doing the review.

Since I have also touched this branch and updated some documentation, we need someone else to review it.

I hope that we will have someone else review this soon.

I have sent it to buildbot

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban removed owner|
|-|-|

Ok. I can no longer review this as I have also made changes to this branch.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl set owner to @hawkowl<br>@hawkowl set status to `closed`|
|-|-|

In changeset 50c5a3a34bbf857d3178758d9012e8e4fe38d6d8
```
#!CommitTicketReference repository="" revision="50c5a3a34bbf857d3178758d9012e8e4fe38d6d8"
Merge 8346-rodrigc-pep3113: Remove use of tuple parameter unpacking

Authors: rodrigc, adiroiban
Reviewers: adiroiban, hawkowl
Fixes: #8346
```

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing review from closed tickets.
